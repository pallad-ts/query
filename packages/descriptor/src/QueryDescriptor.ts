import {
	PaginableByCursor,
	PaginableByOffset,
	Query,
	SortableFieldDefinition,
	SortableSingle,
	Result as _Result,
	SortableMulti,
	ResultMeta,
} from "@pallad/query";
import { Builder } from "@pallad/builder";
import { ERRORS } from "./errors";
import { createResult } from "./utils/createResult";
import { validateSortableFields } from "./utils/validateSortableFields";
import { Either } from "@sweet-monads/either";
import { ValidationViolationError, ViolationsList } from "@pallad/violations";
import { createQueryFactory } from "./utils/createQueryFactory";

function composeDefaultPaginationOptions<
	T extends Partial<Omit<QueryDescriptor.Pagination, "type">>,
>(options?: T) {
	return {
		...QueryDescriptor.defaultPaginationOptions,
		...(options || {}),
	};
}

export class QueryDescriptor<
	TQuery extends Query<any> = Query<unknown>,
	TConfig extends QueryDescriptor.Config = {
		pagination: undefined;
		sorting: undefined;
	},
> extends Builder {
	#config: TConfig;

	#filtersValidation?: QueryDescriptor.FiltersValidator<unknown>;
	#queryFactory?: (input: unknown) => Either<ViolationsList, TQuery>;

	static defaultPaginationOptions: Omit<QueryDescriptor.Pagination, "type"> = {
		defaultLimit: 50,
		maxLimit: 1000,
	};

	constructor() {
		super();

		this.#config = {
			pagination: undefined,
			sorting: undefined,
		} as TConfig;
	}

	cursorPagination(options?: Partial<Omit<QueryDescriptor.Pagination.ByCursor, "type">>) {
		this.#config.pagination = {
			type: "byCursor",
			...composeDefaultPaginationOptions(options),
		};
		this.#reset();
		return this as unknown as QueryDescriptor<
			TQuery & PaginableByCursor,
			Omit<TConfig, "pagination"> & { pagination: QueryDescriptor.Pagination.ByCursor }
		>;
	}

	offsetPagination(options?: Partial<Omit<QueryDescriptor.Pagination.ByOffset, "type">>) {
		this.#config.pagination = {
			type: "byOffset",
			...composeDefaultPaginationOptions(options),
		};
		this.#reset();
		return this as unknown as QueryDescriptor<
			TQuery & PaginableByOffset,
			Omit<TConfig, "pagination"> & { pagination: QueryDescriptor.Pagination.ByOffset }
		>;
	}

	filtersValidator<TFilters>(validation: QueryDescriptor.FiltersValidator<TFilters>) {
		this.#filtersValidation = validation;
		this.#reset();
		return this as QueryDescriptor<Omit<TQuery, "filters"> & Query<TFilters>, TConfig>;
	}

	singleSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: SortableFieldDefinition<TSortableField>
	) {
		validateSortableFields(sortableFields);
		this.#config.sorting = {
			type: "single",
			sortableFields,
			defaultSorting,
		};
		this.#reset();
		return this as unknown as QueryDescriptor<
			TQuery & SortableSingle<TSortableField>,
			Omit<TConfig, "sorting"> & {
				sorting: QueryDescriptor.Sorting.Single<TSortableField>;
			}
		>;
	}

	multiSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: Array<SortableFieldDefinition<TSortableField>>
	) {
		validateSortableFields(sortableFields);
		this.#config.sorting = {
			type: "multi",
			sortableFields,
			defaultSorting,
		};
		this.#reset();
		return this as unknown as QueryDescriptor<
			TQuery & SortableMulti<TSortableField>,
			Omit<TConfig, "sorting"> & {
				sorting: QueryDescriptor.Sorting.Multi<TSortableField>;
			}
		>;
	}

	get sortingConfig() {
		return this.#config.sorting;
	}

	get paginationConfig() {
		return this.#config.pagination;
	}

	#reset() {
		this.#queryFactory = undefined;
	}

	validate() {
		if (this.#config.pagination && this.#config.pagination.type === "byCursor") {
			if (!this.#config.sorting) {
				throw ERRORS.MISSING_SINGLE_SORTING_FOR_CURSOR_PAGINATION.create();
			}

			if (this.#config.sorting.type === "multi") {
				throw ERRORS.MULTI_SORTING_FOR_ALLOWED_FOR_CURSOR_PAGINATION.create();
			}
		}
	}

	createResult<TEntity>(
		options: QueryDescriptor.ResultOptions<TEntity, TQuery, TConfig["pagination"]>
	) {
		this.validate();
		return createResult(this.#config, options) as QueryDescriptor.Result<TEntity, TConfig>;
	}

	createQuery(input: unknown): Either<ViolationsList, TQuery> {
		if (!this.#queryFactory) {
			this.#queryFactory = createQueryFactory(this.#config, this.#filtersValidation) as (
				input: unknown
			) => Either<ViolationsList, TQuery>;
		}

		return this.#queryFactory(input);
	}

	createQueryOrFail(input: unknown, errorMessage?: string): TQuery {
		const result = this.createQuery(input);
		if (result.isLeft()) {
			throw new ValidationViolationError(result.value, errorMessage ?? "Invalid query");
		}
		return result.value;
	}
}

export namespace QueryDescriptor {
	export interface Config {
		pagination: Pagination | undefined;
		sorting: Sorting<any> | undefined;
	}

	export type QueryType<T extends QueryDescriptor<any, any>> =
		T extends QueryDescriptor<infer TQuery, any> ? TQuery : never;

	export type Pagination = Pagination.ByCursor | Pagination.ByOffset;
	export namespace Pagination {
		export interface ByCursor {
			type: "byCursor";
			defaultLimit: number;
			maxLimit: number;
		}

		export interface ByOffset {
			type: "byOffset";
			defaultLimit: number;
			maxLimit: number;
		}
	}

	export type Sorting<TSortableField extends string> =
		| Sorting.Single<TSortableField>
		| Sorting.Multi<TSortableField>;
	export namespace Sorting {
		export interface Single<TSortableField extends string> {
			type: "single";
			sortableFields: readonly [TSortableField, ...TSortableField[]];
			defaultSorting: SortableFieldDefinition<TSortableField>;
		}

		export interface Multi<TSortableField extends string> {
			type: "multi";
			sortableFields: readonly [TSortableField, ...TSortableField[]];
			defaultSorting: Array<SortableFieldDefinition<TSortableField>>;
		}
	}

	export type ResultOptions<
		TEntity,
		TQuery extends Query<any>,
		TPagination extends Pagination | undefined,
	> = ResultOptions.Base<TEntity, TQuery> &
		(TPagination extends Pagination.ByCursor
			? {
					nextPageCursor?: string;
					previousPageCursor?: string;
				}
			: {});
	export namespace ResultOptions {
		export interface Base<TEntity, TQuery extends Query<any>> {
			results: TEntity[];
			query: TQuery;
		}
	}

	export type ResultMetaType<T extends QueryDescriptor<any, any>> =
		ReturnType<T["createResult"]> extends { meta: infer TMeta } ? TMeta : never;

	export type Result<TEntity, TConfig extends Config> = _Result<TEntity> & Result.Meta<TConfig>;

	export namespace Result {
		export type Meta<T extends Config> = T extends { pagination: undefined; sorting: undefined }
			? {}
			: ResultMeta<MetaPagination<T["pagination"]> & MetaSorting<T["sorting"]>>;

		export type MetaPagination<T extends Pagination | undefined> = T extends undefined
			? {}
			: T extends Pagination.ByCursor
				? PaginableByCursor.ResultMeta
				: PaginableByOffset.ResultMeta;

		export type MetaSorting<T extends Sorting<any> | undefined> = T extends undefined
			? {}
			: T extends Sorting.Single<infer TField>
				? SortableSingle.ResultMeta<TField>
				: T extends Sorting.Multi<infer TField>
					? SortableMulti.ResultMeta<TField>
					: {};
	}

	export type FiltersValidator<TFilters> = (input: unknown) => Either<ViolationsList, TFilters>;
}
