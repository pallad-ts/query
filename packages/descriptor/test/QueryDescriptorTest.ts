import { QueryDescriptor } from "@src/QueryDescriptor";
import { assert, IsExact } from "conditional-type-checks";
import {
	PaginableByCursor,
	PaginableByOffset,
	Query,
	Result,
	ResultMeta,
	SortableMulti,
	SortableSingle,
} from "@pallad/query";
import { DEFAULT_SORTING, SORTABLE_FIELDS } from "./fixtures";
import { Either, right } from "@sweet-monads/either";
import { ViolationsList } from "@pallad/violations";
import { fromZodValidation } from "@pallad/violations-zod";
import { z } from "zod";

describe("QueryDescriptor", () => {
	describe("pagination", () => {
		it("by default not set", () => {
			const descriptor = new QueryDescriptor();
			expect(descriptor.paginationConfig).toBeUndefined();
		});

		describe("by cursor", () => {
			it("types", () => {
				const descriptor = new QueryDescriptor().cursorPagination();
				type ExpectedQuery = Query<unknown> & PaginableByCursor;
				type ExpectedResultMeta = PaginableByCursor.ResultMeta;

				assert<IsExact<QueryDescriptor.QueryType<typeof descriptor>, ExpectedQuery>>(true);
				assert<
					IsExact<QueryDescriptor.ResultMetaType<typeof descriptor>, ExpectedResultMeta>
				>(true);
			});

			it("applies default config if not set explicitly", () => {
				const config = new QueryDescriptor().cursorPagination();

				expect(config.paginationConfig).toEqual({
					type: "byCursor",
					defaultLimit: 50,
					maxLimit: 1000,
				});
			});

			it("merges with default settings", () => {
				const config = new QueryDescriptor().cursorPagination({ defaultLimit: 10 });

				expect(config.paginationConfig).toEqual({
					type: "byCursor",
					defaultLimit: 10,
					maxLimit: 1000,
				});
			});
		});

		describe("by offset", () => {
			it("types", () => {
				const descriptor = new QueryDescriptor().offsetPagination();
				type ExpectedQuery = Query<unknown> & PaginableByOffset;
				type ExpectedResultMeta = PaginableByOffset.ResultMeta;

				assert<IsExact<QueryDescriptor.QueryType<typeof descriptor>, ExpectedQuery>>(true);
				assert<
					IsExact<QueryDescriptor.ResultMetaType<typeof descriptor>, ExpectedResultMeta>
				>(true);
			});

			it("applies default config if not set explicitly", () => {
				const descriptor = new QueryDescriptor().offsetPagination();

				expect(descriptor.paginationConfig).toEqual({
					type: "byOffset",
					defaultLimit: 50,
					maxLimit: 1000,
				});
			});

			it("merges with default settings", () => {
				const descriptor = new QueryDescriptor().offsetPagination({ defaultLimit: 10 });

				expect(descriptor.paginationConfig).toEqual({
					type: "byOffset",
					defaultLimit: 10,
					maxLimit: 1000,
				});
			});
		});
	});

	describe("sorting", () => {
		it("by default not set", () => {
			const config = new QueryDescriptor();
			expect(config.sortingConfig).toBeUndefined();
		});

		describe("single", () => {
			const descriptor = new QueryDescriptor().singleSorting(["foo", "bar"], {
				field: "foo",
				direction: "ASC",
			});

			it("types", () => {
				type ExpectedQuery = Query<unknown> & SortableSingle<"foo" | "bar">;
				type ExpectedResultMeta = SortableSingle.ResultMeta<"foo" | "bar">;

				assert<IsExact<QueryDescriptor.QueryType<typeof descriptor>, ExpectedQuery>>(true);
				assert<
					IsExact<QueryDescriptor.ResultMetaType<typeof descriptor>, ExpectedResultMeta>
				>(true);
			});

			it("getting config", () => {
				expect(descriptor.sortingConfig).toEqual({
					type: "single",
					sortableFields: ["foo", "bar"],
					defaultSorting: { field: "foo", direction: "ASC" },
				});
			});
		});

		describe("multi", () => {
			const descriptor = new QueryDescriptor().multiSorting(
				["foo", "bar"],
				[
					{ field: "foo", direction: "ASC" },
					{ field: "bar", direction: "DESC" },
				]
			);

			it("types", () => {
				type ExpectedQuery = Query<unknown> & SortableMulti<"foo" | "bar">;
				type ExpectedResultMeta = SortableMulti.ResultMeta<"foo" | "bar">;

				assert<IsExact<QueryDescriptor.QueryType<typeof descriptor>, ExpectedQuery>>(true);
				assert<
					IsExact<QueryDescriptor.ResultMetaType<typeof descriptor>, ExpectedResultMeta>
				>(true);
			});

			it("getting config", () => {
				expect(descriptor.sortingConfig).toEqual({
					type: "multi",
					sortableFields: ["foo", "bar"],
					defaultSorting: [
						{ field: "foo", direction: "ASC" },
						{ field: "bar", direction: "DESC" },
					],
				});
			});
		});
	});

	describe("creating query", () => {
		describe("filters", () => {
			it("types", () => {
				const descriptor = new QueryDescriptor().filtersValidator(() =>
					right({
						foo: "bar",
					} as const)
				);

				type Expected = Query<{ foo: "bar" }>;
				assert<IsExact<QueryDescriptor.QueryType<typeof descriptor>, Expected>>(true);
			});

			it("by default accepts all filters", () => {
				const descriptor = new QueryDescriptor();

				expect(descriptor.createQuery({ filters: { literally: "anything" } })).toEqual(
					right({
						filters: { literally: "anything" },
					})
				);
			});

			describe("uses filters schema if provided", () => {
				const builder = new QueryDescriptor().filtersValidator(input => {
					return fromZodValidation(
						z.object({
							foo: z.literal("foo"),
							bar: z.literal("bar").optional(),
						}),
						input
					);
				});

				it("success", () => {
					expect(
						builder.createQuery({
							filters: {
								foo: "foo",
								bar: "bar",
							},
						})
					).toEqual(
						right({
							filters: { foo: "foo", bar: "bar" },
						})
					);
				});

				it("failure", async () => {
					const result = await builder.createQuery({ filters: {} });

					expect(result.isLeft()).toBe(true);

					expect(result.value).toMatchSnapshot();
				});
			});
		});

		describe("pagination", () => {
			describe.each<[string, QueryDescriptor<any, any>]>([
				[
					"by cursor",
					new QueryDescriptor()
						.cursorPagination()
						.singleSorting(["foo", "bar"], { field: "foo", direction: "ASC" }),
				],
				["by offset", new QueryDescriptor().offsetPagination()],
			])("common (for %s)", (name, qb) => {
				it("limit cannot be lower than 1", async () => {
					const result = await qb.createQuery({ limit: 0 });
					expect(result.value).toMatchSnapshot();
				});

				it("uses default limit", async () => {
					const result = await qb.createQuery({});
					expect(result.value).toMatchObject({ limit: 50 });
				});

				it("limit cannot be greater than max limit", async () => {
					const result = await qb.createQuery({ limit: 5000 });
					expect(result.value).toMatchSnapshot();
				});

				it("limit must be integer", async () => {
					const result = await qb.createQuery({ limit: "asdfa" });
					expect(result.value).toMatchSnapshot();
				});
			});

			describe("by cursor", () => {
				const descriptor = new QueryDescriptor()
					.cursorPagination()
					.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING);

				it("types", () => {
					type Expected = Either<
						ViolationsList,
						Query<unknown> & PaginableByCursor & SortableSingle<"foo" | "bar">
					>;

					assert<IsExact<ReturnType<(typeof descriptor)["createQuery"]>, Expected>>(true);
				});

				describe.each([["after"], ["before"]])('allows "%s" cursor property', property => {
					it("fails if not a string", () => {
						const result = descriptor.createQuery({ [property]: [] });
						expect(result.value).toMatchSnapshot();
					});

					it("success", () => {
						const value = `${property} cursor`;
						expect(descriptor.createQuery({ [property]: value })).toEqual(
							right({
								filters: {},
								[property]: value,
								limit: 50,
								sortBy: DEFAULT_SORTING,
							})
						);
					});
				});
			});

			describe("by offset", () => {
				const descriptor = new QueryDescriptor().offsetPagination();

				it("types", () => {
					type Expected = Either<ViolationsList, Query<unknown> & PaginableByOffset>;
					assert<IsExact<ReturnType<(typeof descriptor)["createQuery"]>, Expected>>(true);
				});

				it("offset by default is 0", () => {
					expect(descriptor.createQuery({})).toEqual(
						right({
							filters: {},
							limit: 50,
							offset: 0,
						})
					);
				});

				it("offset cannot be lower than 0", () => {
					const result = descriptor.createQuery({ offset: -1 });
					expect(result.value).toMatchSnapshot();
				});

				it("offset must be an integer", () => {
					const result = descriptor.createQuery({ offset: "asdfasdf" });
					expect(result.value).toMatchSnapshot();
				});
			});
		});

		describe("sorting", () => {
			describe.each([
				[
					"single",
					new QueryDescriptor().singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING),
				],
				[
					"multi",
					new QueryDescriptor().multiSorting(SORTABLE_FIELDS.slice(), [DEFAULT_SORTING]),
				],
			])("common (for %s)", (name, qb) => {
				const isMulti = name === "multi";

				function testCase(sorting: any) {
					return qb.createQuery({ sortBy: isMulti ? [sorting] : sorting });
				}

				it("forces to use only allowed sortable fields", async () => {
					const result = await testCase({ field: "unsupported", direction: "ASC" });

					expect(result.value).toMatchSnapshot();
				});

				it("sortable field is required", async () => {
					const result = await testCase({ direction: "ASC" });

					expect(result.value).toMatchSnapshot();
				});

				it("sorting direction is required", async () => {
					const result = await testCase({ field: "foo" });

					expect(result.value).toMatchSnapshot();
				});

				it("fails for invalid sorting direction", async () => {
					const result = await testCase({ field: "foo", direction: "bas" });

					expect(result.value).toMatchSnapshot();
				});

				it.each([["ASC"], ["DESC"]])("sorting direction must be ASC or DESC", direction => {
					let sorting = { field: "foo", direction };
					expect(testCase(sorting)).toEqual(
						right({
							filters: {},
							sortBy: isMulti ? [sorting] : sorting,
						})
					);
				});

				it("applies default sorting if not explicitly set", () => {
					expect(qb.createQuery({})).toEqual(
						right({
							filters: {},
							sortBy: isMulti ? [DEFAULT_SORTING] : DEFAULT_SORTING,
						})
					);
				});
			});

			describe("single", () => {
				const descriptor = new QueryDescriptor().singleSorting(
					SORTABLE_FIELDS.slice(),
					DEFAULT_SORTING
				);

				it("types", () => {
					type Expected = Either<
						ViolationsList,
						Query<unknown> & SortableSingle<"foo" | "bar">
					>;

					assert<IsExact<ReturnType<(typeof descriptor)["createQuery"]>, Expected>>(true);
				});
			});

			describe("multi", () => {
				const builder = new QueryDescriptor().multiSorting(SORTABLE_FIELDS.slice(), [
					DEFAULT_SORTING,
				]);

				it("types", () => {
					type Expected = Either<
						ViolationsList,
						Query<unknown> & SortableMulti<"foo" | "bar">
					>;
					assert<IsExact<ReturnType<(typeof builder)["createQuery"]>, Expected>>(true);
				});

				it("sorting array cannot be empty", () => {
					const result = builder.createQuery({
						sortBy: [],
					});
					expect(result.value).toMatchSnapshot();
				});
			});
		});
	});

	// describe("query builder", () => {
	// 	it("required config to be valid before returning query builder", () => {
	// 		expect(() => {
	// 			new QueryConfig().cursorPagination().getQueryBuilder();
	// 		}).toThrowErrorMatchingSnapshot();
	// 	});
	//
	// 	it("types", () => {
	// 		const builder = new QueryConfig()
	// 			.singleSorting(["foo", "bar"], { field: "bar", direction: "ASC" })
	// 			.filters<Filters>()
	// 			.getQueryBuilder();
	//
	// 		type Expected = QueryBuilder<QueryDescriptor<Filters, SortableSingle<"foo" | "bar">>>;
	// 		assert<IsExact<typeof builder, Expected>>(true);
	// 	});
	// });

	describe("building result", () => {
		describe("none of options selected", () => {
			const descriptor = new QueryDescriptor();
			it("types", () => {
				type Expected = Result<unknown>;
				type Input = ReturnType<(typeof descriptor)["createResult"]>;

				const o: Input = { results: [] };
				assert<IsExact<Input, Expected>>(true);
			});

			it("building result", () => {
				expect(
					descriptor.createResult({
						results: [1, 2],
						query: {
							filters: {},
						},
					})
				).toEqual({
					results: [1, 2],
				});
			});
		});

		describe("cursor pagination", () => {
			const descriptor = new QueryDescriptor()
				.cursorPagination()
				.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING);

			it("types", () => {
				type Expected = Result<unknown> &
					ResultMeta<
						PaginableByCursor.ResultMeta & SortableSingle.ResultMeta<"foo" | "bar">
					>;
				type Input = ReturnType<(typeof descriptor)["createResult"]>;

				assert<IsExact<Input, Expected>>(true);
			});

			it("building result", () => {
				expect(
					descriptor.createResult({
						results: [1, 2],
						query: {
							filters: {},
							limit: 10,
							sortBy: DEFAULT_SORTING,
							after: "after",
							before: "before",
						},
						nextPageCursor: "nextPage",
						previousPageCursor: "previousPage",
					})
				).toEqual({
					results: [1, 2],
					meta: {
						limit: 10,
						nextPage: "nextPage",
						previousPage: "previousPage",
						sortBy: DEFAULT_SORTING,
					},
				});
			});
		});

		describe("offset pagination", () => {
			const descriptor = new QueryDescriptor().offsetPagination();

			it("types", () => {
				type Expected = Result<unknown> & ResultMeta<PaginableByOffset.ResultMeta>;
				type Input = ReturnType<(typeof descriptor)["createResult"]>;

				assert<IsExact<Input, Expected>>(true);
			});

			it("building result", () => {
				expect(
					descriptor.createResult({
						results: [1, 2],
						query: {
							filters: {},
							limit: 10,
							offset: 10,
						},
					})
				).toEqual({
					results: [1, 2],
					meta: {
						limit: 10,
						offset: 10,
					},
				});
			});
		});

		describe("multi sorting", () => {
			const descriptor = new QueryDescriptor().multiSorting(SORTABLE_FIELDS.slice(), [
				DEFAULT_SORTING,
			]);

			it("types", () => {
				type Expected = Result<unknown> &
					ResultMeta<SortableMulti.ResultMeta<"foo" | "bar">>;
				type Input = ReturnType<(typeof descriptor)["createResult"]>;

				assert<IsExact<Input, Expected>>(true);
			});

			it("building result", () => {
				expect(
					descriptor.createResult({
						results: [1, 2],
						query: {
							filters: {},
							sortBy: [
								{ field: "foo", direction: "ASC" },
								{ field: "bar", direction: "DESC" },
							],
						},
					})
				).toEqual({
					results: [1, 2],
					meta: {
						sortBy: [
							{ field: "foo", direction: "ASC" },
							{ field: "bar", direction: "DESC" },
						],
					},
				});
			});
		});
	});

	describe("validation", () => {
		it("pagination by cursor cannot be used without single sorting setup", () => {
			expect(() => {
				new QueryDescriptor().cursorPagination().validate();
			}).toThrowErrorMatchingSnapshot();
		});

		it("pagination by cursor cannot be used with multi sorting", () => {
			expect(() => {
				new QueryDescriptor()
					.cursorPagination()
					.multiSorting(["foo", "bar"], [{ field: "foo", direction: "ASC" }])
					.validate();
			}).toThrowErrorMatchingSnapshot();
		});
	});
});
