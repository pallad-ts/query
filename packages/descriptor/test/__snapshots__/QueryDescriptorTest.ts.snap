// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryDescriptor creating query filters uses filters schema if provided failure 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Invalid literal value, expected "foo"",
      "params": undefined,
      "path": [
        "foo",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination by cursor allows "after" cursor property fails if not a string 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Expected string, received array",
      "params": undefined,
      "path": [
        "after",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination by cursor allows "before" cursor property fails if not a string 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Expected string, received array",
      "params": undefined,
      "path": [
        "before",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination by offset offset cannot be lower than 0 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Number must be greater than or equal to 0",
      "params": undefined,
      "path": [
        "offset",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination by offset offset must be an integer 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Expected number, received string",
      "params": undefined,
      "path": [
        "offset",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by cursor) limit cannot be greater than max limit 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Number must be less than or equal to 1000",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by cursor) limit cannot be lower than 1 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Number must be greater than or equal to 1",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by cursor) limit must be integer 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Expected number, received string",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by offset) limit cannot be greater than max limit 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Number must be less than or equal to 1000",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by offset) limit cannot be lower than 1 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Number must be greater than or equal to 1",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query pagination common (for by offset) limit must be integer 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Expected number, received string",
      "params": undefined,
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for multi) fails for invalid sorting direction 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Invalid enum value. Expected 'DESC' | 'ASC', received 'bas'",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for multi) forces to use only allowed sortable fields 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Invalid enum value. Expected 'foo' | 'bar', received 'unsupported'",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for multi) sortable field is required 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Required",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for multi) sorting direction is required 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Required",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for single) fails for invalid sorting direction 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Invalid enum value. Expected 'DESC' | 'ASC', received 'bas'",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for single) forces to use only allowed sortable fields 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Invalid enum value. Expected 'foo' | 'bar', received 'unsupported'",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for single) sortable field is required 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Required",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting common (for single) sorting direction is required 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Required",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor creating query sorting multi sorting array cannot be empty 1`] = `
ViolationsList {
  "storage": [
    {
      "code": undefined,
      "message": "Array must contain at least 1 element(s)",
      "params": undefined,
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;

exports[`QueryDescriptor validation pagination by cursor cannot be used with multi sorting 1`] = `"Sorting by multiple fields is not allowed in pagination by cursor"`;

exports[`QueryDescriptor validation pagination by cursor cannot be used without single sorting setup 1`] = `"Pagination by cursor cannot be used without single field sorting"`;
